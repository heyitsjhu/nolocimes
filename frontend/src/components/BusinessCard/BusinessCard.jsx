import React, { useContext } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Backdrop from '@material-ui/core/Backdrop';
import Box from '@material-ui/core/Box';
import Link from '@material-ui/core/Link';
import Fade from '@material-ui/core/Fade';
import Modal from '@material-ui/core/Modal';
import Typography from '@material-ui/core/Typography';
import DomainIcon from '@material-ui/icons/Domain';
import EmailIcon from '@material-ui/icons/Email';
import GitHubIcon from '@material-ui/icons/GitHub';
import LinkedInIcon from '@material-ui/icons/LinkedIn';
import PersonIcon from '@material-ui/icons/Person';

import IconButton from 'components/IconButton/IconButton';
import SiteLogo from 'components/SiteLogo/SiteLogo';
import { LINKS, STORE_KEYS } from 'const';
import { useCopy } from 'hooks/useCopy';
import { AppContext } from 'stores';
import { updateAppState } from 'stores/actions/appActions';

const useStyles = makeStyles(({ breakpoints, palette, shared, spacing, zIndex }) => ({
  businessCardLayout: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-start',
    margin: '10vh auto',
    padding: spacing(3),
    height: '100%',
    width: '100%',
    maxWidth: '44vh',
    maxHeight: '66vh',
    borderRadius: 0,
    border: shared.borderDefault,
    borderTop: `.3125rem solid ${palette.primary.main}`,
    backgroundColor: palette.grey[1300],
    boxShadow: '0.125rem 0.375rem 0.5rem 0 rgba(0,0,0,0.6)',
    outline: 'none',
    zIndex: zIndex.businessCard,
    [`${breakpoints.down('sm')} and (orientation: landscape)`]: {
      maxWidth: '66vw',
    },
  },
  backdrop: {
    backgroundColor: 'rgba(0,0,0,0.75)',
  },
  logoTitleContainer: {
    [`${breakpoints.down('sm')} and (orientation: landscape)`]: {
      display: 'flex',
      alignItems: 'flex-end',
    },
  },
  contentContainer: {
    flex: 'none',
    marginTop: 'auto',
    padding: 0,
  },
  siteLogo: {
    marginTop: spacing(1),
    marginBottom: spacing(2),
    '& path': {
      '&:nth-child(6),&:nth-child(7),&:nth-child(8)': {
        fill: palette.primary.main,
      },
    },
  },
  titleContainer: {
    marginTop: spacing(1),
    marginBottom: 'auto',
    padding: 0,
    [`${breakpoints.down('sm')} and (orientation: landscape)`]: {
      marginLeft: spacing(3),
      marginTop: spacing(2),
    },
  },
  title: {
    color: palette.primary.main,
    letterSpacing: '.05rem',
  },
  contentIcon: {
    paddingRight: spacing(1),
  },
  contentText: {
    display: 'flex',
    alignItems: 'center',
    color: palette.grey[300],
    letterSpacing: '.05rem',
    lineHeight: 1.65,
  },
  name: {
    color: palette.grey[300],
    textTransform: 'uppercase',
    letterSpacing: '0.2rem',
    lineHeight: 1.1,
  },
}));

export default () => {
  const { t } = useCopy();
  const classes = useStyles();
  const [appState, dispatch] = useContext(AppContext);
  const { navControlsEnabled } = appState[STORE_KEYS.SITE_SETTINGS];

  const openBusinessCard = () => dispatch(updateAppState(STORE_KEYS.BUSINESS_CARD, 'show', true));
  const closeBusinessCard = () => dispatch(updateAppState(STORE_KEYS.BUSINESS_CARD, 'show', false));

  return (
    <>
      <IconButton
        aria-label={t('a11y.ariaLabel.businessCardButton')}
        onClick={navControlsEnabled ? openBusinessCard : undefined}
      >
        <PersonIcon fontSize="small" />
      </IconButton>
      <Modal
        // need to associate with content in business card
        aria-labelledby="business-card-modal-title"
        aria-describedby="business-card-modal-description"
        open={appState[STORE_KEYS.BUSINESS_CARD].show}
        onClose={closeBusinessCard}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 500,
          className: classes.backdrop,
        }}
      >
        <Fade in={appState[STORE_KEYS.BUSINESS_CARD].show}>
          <Box className={classes.businessCardLayout}>
            <Box className={classes.logoTitleContainer}>
              <SiteLogo className={classes.siteLogo} size={100} />
              <Box className={classes.titleContainer}>
                <Typography className={classes.name} component="h2" variant="h2">
                  {t('components.BusinessCard.name')}
                </Typography>
                <Typography className={classes.title} component="span" variant="body2">
                  {t('components.BusinessCard.title')}
                </Typography>
                <Box marginLeft={-0.5}>
                  <IconButton
                    aria-label={t('a11y.ariaLabel.githubUserUrl')}
                    href={LINKS.GITHUB_USER_URL}
                    target="_blank"
                  >
                    <GitHubIcon fontSize="small" />
                  </IconButton>

                  <IconButton
                    aria-label={t('a11y.ariaLabel.linkedInUrl')}
                    href={LINKS.LINKEDIN_URL}
                    target="_blank"
                  >
                    <LinkedInIcon />
                  </IconButton>
                </Box>
              </Box>
            </Box>
            <Box className={classes.contentContainer}>
              <Link
                aria-label={t('a11y.ariaLabel.businessCardEmail')}
                href={`mailto:${t('components.BusinessCard.emailAddress')}`}
              >
                <Typography className={classes.contentText} component="span" variant="body2">
                  <EmailIcon className={classes.contentIcon} />
                  {t('components.BusinessCard.emailAddress')}
                </Typography>
              </Link>
              <Typography className={classes.contentText} component="span" variant="body2">
                <DomainIcon className={classes.contentIcon} />
                {t('components.BusinessCard.companyName')}
              </Typography>
            </Box>
          </Box>
        </Fade>
      </Modal>
    </>
  );
};
