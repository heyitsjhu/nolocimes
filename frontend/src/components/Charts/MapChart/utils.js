export const handleOnCountryOver = (event, imageSeries) => {
  rollOverCountry(event.target, imageSeries);
};

export const handleOnCountryOut = (event, imageSeries) => {
  rollOutCountry(event.target, imageSeries);
};

export const handleOnImageOver = (event, polygonSeries) => {
  rollOverCountry(
    polygonSeries.getPolygonById(event.target.dataItem.dataContext.id),
    event.target.parent
  );
};

export const handleOnImageOut = (event, polygonSeries) => {
  rollOutCountry(
    polygonSeries.getPolygonById(event.target.dataItem.dataContext.id),
    event.target.parent
  );
};

// what happens when a country is rolled-over
/**
 *
 * @param {*} mapPolygon
 * @param {*} imageSeries
 */
const rollOverCountry = (mapPolygon, imageSeries) => {
  // resetHover(mapPolygon.parent, imageSeries);

  if (mapPolygon) {
    mapPolygon.isHover = true;

    // make bubble hovered too
    const image = imageSeries.getImageById(mapPolygon.dataItem.dataContext.id);

    if (image) {
      image.dataItem.dataContext.name = mapPolygon.dataItem.dataContext.name;
      image.isHover = true;
      image.dataItem.dataContext.isHover = true;
    }
  }
};

const rollOutCountry = (mapPolygon, imageSeries) => {
  if (!mapPolygon) return;

  var image = imageSeries.getImageById(mapPolygon.dataItem.dataContext.id);

  resetHover(mapPolygon.parent, imageSeries);
  if (image) {
    image.isHover = false;
  }
};

const resetHover = (mapPolygon, imageSeries) => {
  mapPolygon.mapPolygons.each(polygon => {
    polygon.isHover = false;
  });

  imageSeries.mapImages.each(image => {
    image.isHover = false;
  });
};

/**
 * Positions the tooltip to the top of the circle (bubble) image by
 * offsetting its Y-position by the radius of the circle.
 * @param {*} tooltipY
 * @param {*} target - imageTemplate [amcharts]
 * @returns {number}
 */
export const positionTooltipToTop = (tooltipY, target) => {
  return -target.children.getIndex(0).radius;
};

/**
 *
 * @param {*} positionType
 * @param {*} positionData
 * @param {*} target
 * @param {*} polygonSeries
 * @returns {object}
 */
export const centerImageBubbleOverCountry = (positionType, positionData, target, polygonSeries) => {
  const { dataContext } = target.dataItem;
  var polygon = polygonSeries.getPolygonById(dataContext.id);
  if (polygon) return polygon[`visual${positionType}`];

  return positionData;
};

export const renderHTMLTooltip = (metric, options, t) => {
  return `
  <div class="MuiBox-root">
    <strong>{name}</strong> <br />
    <p class="MuiTypography-root MuiTypography-caption">
    ${t('common.population')}:
    <span class="MuiTypography-root MuiTypography-overline">{population}</span>
    </p>
    <hr />
    <span class="MuiTypography-root MuiTypography-overline">{day}</span>
    ${options
      .map(
        option => `
      <p class="MuiTypography-root MuiTypography-caption">
        ${option === metric ? '<strong>' : ''}
          ${t('pages.Coronavirus.labels.' + option)}: 
        ${option === metric ? '</strong>' : ''}
        <span class="MuiTypography-root MuiTypography-${
          option === metric ? 'caption' : 'overline'
        }">
          ${option === metric ? '<strong>' : ''}
          {${option}}
          ${option === metric ? '</strong>' : ''}
        </span>
      </p>
    `
      )
      .join('')}
  </div>
  `;
};
