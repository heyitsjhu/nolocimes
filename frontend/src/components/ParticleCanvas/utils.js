export const getRandomSpeed = pos => {
  const [min, max] = [-1, 1];
  switch (pos) {
    case 'top':
      return [randomNumFrom(min, max), randomNumFrom(0.1, max)];
    case 'right':
      return [randomNumFrom(min, -0.1), randomNumFrom(min, max)];
    case 'bottom':
      return [randomNumFrom(min, max), randomNumFrom(min, -0.1)];
    case 'left':
      return [randomNumFrom(0.1, max), randomNumFrom(min, max)];
    default:
      return;
  }
};
export const randomArrayItem = arr => {
  return arr[Math.floor(Math.random() * arr.length)];
};
export const randomNumFrom = (min, max) => {
  return Math.random() * (max - min) + min;
};

export const randomSidePos = length => {
  return Math.ceil(Math.random() * length);
};

export const getDisOf = (b1, b2) => {
  const delta_x = Math.abs(b1.x - b2.x);
  const delta_y = Math.abs(b1.y - b2.y);

  return Math.sqrt(delta_x * delta_x + delta_y * delta_y);
};
