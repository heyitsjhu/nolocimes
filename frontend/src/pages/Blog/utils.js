export const stateReducer = (state, action) => {
  if (action.key === 'post') {
    return { ...state, postContent: action.post };
  }

  const { index } = state;
  const prevIndexState = {
    prev: index.prev > 0 ? index.prev - 1 : null,
    current: index.prev,
    next: index.current,
  };
  const nextIndexState = {
    prev: index.current,
    current: index.next,
    next: index.next + 1,
  };
  const indexConfig = action.key === 'prev' ? prevIndexState : nextIndexState;

  return {
    ...state,
    index: indexConfig,
  };
};
