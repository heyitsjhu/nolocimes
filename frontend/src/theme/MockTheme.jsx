import React from 'react';
import { createMuiTheme } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';

const MockTheme = ({ children }) => {
  const theme = createMuiTheme({ shared: {} });
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export default MockTheme;
