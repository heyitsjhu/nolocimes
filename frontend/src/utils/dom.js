import DLError from 'models/DLError';
import { capitalize } from './';

/**
 *
 * @param {*} type
 * @param {*} name
 */
export const getElClass = (type, name) => {
  if (!name) {
    throw new DLError('Name parameter cannot be empty');
  }

  type = type ? capitalize(type) : '';
  return `DL${type}-${name}`;
};

/**
 *
 * @param {*} type
 * @param {*} id
 */
export const getElId = (type, name) => {
  type = type ? `${type}-` : '';
  return `dlabs-${type}${name || Math.round(Math.random() * 100000)}`;
};

// export const preventDefault = event => event.preventDefault();
