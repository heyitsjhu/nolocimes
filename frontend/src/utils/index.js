export * from './dom';
export * from './formatters';
export * from './postHelpers';
export * from './rfdc';

/**
 *
 * @param {*} str
 */
export const capitalize = str => {
  return str ? str.charAt(0).toUpperCase() + str.slice(1) : '';
};
