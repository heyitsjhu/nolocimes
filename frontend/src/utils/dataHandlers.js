import { deepClone } from 'utils';

export const COVID_CHART_DATA_LABELS = [
  { key: 'cases_new', label: 'New Cases' },
  { key: 'cases_active', label: 'Active Cases' },
  { key: 'cases_critical', label: 'Critical Cases' },
  { key: 'cases_recovered', label: 'Recovered Cases' },
  { key: 'cases_total', label: 'Total Cases' },
  { key: 'deaths_new', label: 'New Deaths' },
  { key: 'deaths_total', label: 'Total Deaths' },
  { key: 'tests_total', label: 'Total Tests' },
];

// Merges Covid-19 historical data for an array of countries into the
// reducer's coronavirus history object for amCharts ingestion. Also, pushes
// the country into a _retrievedCountries property for easier reference
export const convertCovidHistoricalData = (existingData, countryData) => {
  const newState = existingData ? deepClone(existingData) : {};
  const stateKeys = Object.keys(existingData).filter(key => key !== '_retrievedCountries');

  if (newState._retrievedCountries.includes(countryData[0].country)) return;
  if (countryData.length > 0) {
    // just tracks which countries we've fetched data for already.
    newState._retrievedCountries.push(countryData[0].country);
  }

  countryData.reduce(getEndOfDayStatistic, []).forEach(dataPoint => {
    const country = dataPoint.country;

    stateKeys.forEach(stateKey => {
      const keys = stateKey.split(/_(.+)/);
      // const daysData = newState[stateKey].find(({ day }) => day === metadata.day);
      const existingDataPointIdx = newState[stateKey].findIndex(d => d.day === dataPoint.day);
      if (existingDataPointIdx > -1) {
        newState[stateKey][existingDataPointIdx][country] =
          stateKey === 'population' ? dataPoint[stateKey] : dataPoint[keys[0]][keys[1]];
      } else {
        newState[stateKey].push({
          [country]: stateKey === 'population' ? dataPoint[stateKey] : dataPoint[keys[0]][keys[1]],
          day: dataPoint.day,
          time: dataPoint.time,
        });
      }
    });
  });

  return newState;
};

// Given a country's historical data set, pushes the end of day, or latest, data for each day into, and returns, the accumulated array.
const getEndOfDayStatistic = (accItem, currItem) => {
  const hasEntry = accItem.find(item => item.day === currItem.day);
  if (!accItem.length || !hasEntry) {
    accItem.push(currItem);
  }
  return accItem;
};

// Converts Covid-19 statistics data for amCharts ingestion.
export const convertCovidStatisticsData = data => {
  const replaceRegex = new RegExp(/-/, 'gi');

  return data.statistics
    .map(({ country, cases, deaths, tests, day, time }) => {
      const countryName = country.replace(replaceRegex, ' ');
      const countryData = data.countries.find(countryItem => countryItem.name === countryName);

      return {
        id: countryData ? countryData.alpha2code : undefined,
        name: countryName,
        cases_new: cases.new,
        cases_active: cases.active,
        cases_critical: cases.critical,
        cases_recovered: cases.recovered,
        cases_total: cases.total,
        deaths_new: deaths.new,
        deaths_total: deaths.total,
        tests_total: tests.total,
        day,
        time,
      };
    })
    .filter(dataItem => typeof dataItem.id !== 'undefined');
};
