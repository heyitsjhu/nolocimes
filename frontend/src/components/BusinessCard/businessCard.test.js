import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { render, fireEvent, screen } from '@testing-library/react';
import {
  getByLabelText,
  getByText,
  getByTestId,
  queryByTestId,
  waitFor,
} from '@testing-library/dom';

import MockTheme from 'theme/MockTheme';
import BusinessCard from './BusinessCard';
import { getTestElId } from 'utils';

jest.mock('const', () => ({
  LINKS: { GITHUB_USER_URL: '' },
  TEST_IDS: { BUSINESS_CARD_MODAL: '' },
  SITE_NAVIGATION: { mapping: [{}, {}, {}, {}, {}, {}, {}, {}] },
  STORE_KEYS: { BLOG: '' },
}));

describe('[components] - BusinessCard', () => {
  let container = null;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });
  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  test.skip('it renders without failing snapshot', async () => {
    const onCloseMock = jest.fn();
    const isOpen = true;
    render(
      <MockTheme>
        <BusinessCard open={isOpen} onClose={onCloseMock} />
      </MockTheme>,
      container
    );
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByRole('presentation'), { button: 1 });
    expect(onCloseMock).toHaveBeenCalled();
  });
});
