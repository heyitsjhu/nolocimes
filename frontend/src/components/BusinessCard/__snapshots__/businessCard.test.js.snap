// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - BusinessCard it renders without failing snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
